pkg/appstate/daemonsets.go

```diff
@@ -0,0 +1,180 @@
+			return clientset.AppsV1().DaemonSets(targetNamespace).List(context.TODO(), options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return clientset.AppsV1().DaemonSets(targetNamespace).Watch(context.TODO(), options)
+		},
+	}
+
+	informer := cache.NewSharedInformer(listwatch, &appsv1.DaemonSet{}, time.Minute)
+
+	eventHandler := &daemonSetEventHandler{
+		informers:       filterStatusInformersByResourceKind(informers, DaemonSetResourceKind),
+		resourceStateCh: resourceStateCh,
+		clientset:       clientset,
+		targetNamespace: targetNamespace,
+	}
+
+	runInformer(ctx, informer, eventHandler)
+}
+
+func (h *daemonSetEventHandler) ObjectCreated(obj interface{}) {
+	r := h.cast(obj)
+	if _, ok := h.getInformer(r); !ok {
+		return
+	}
+
+	h.resourceStateCh <- makeDaemonSetResourceState(r, h.calculateDaemonSetState(r))
+}
+
+func (h *daemonSetEventHandler) ObjectDeleted(obj interface{}) {
+	r := h.cast(obj)
+	if _, ok := h.getInformer(r); !ok {
+		return
+	}
+
+	h.resourceStateCh <- makeDaemonSetResourceState(r, types.StateMissing)
+}
+
+func (h *daemonSetEventHandler) ObjectUpdated(obj interface{}) {
+	r := h.cast(obj)
+	if _, ok := h.getInformer(r); !ok {
+		return
+	}
+
+	h.resourceStateCh <- makeDaemonSetResourceState(r, h.calculateDaemonSetState(r))
+}
+
+func (h *daemonSetEventHandler) getInformer(r *appsv1.DaemonSet) (types.StatusInformer, bool) {
+	if r != nil {
+		for _, informer := range h.informers {
+			if r.Namespace == informer.Namespace && r.Name == informer.Name {
+				return informer, true
+			}
+		}
+	}
+
+	return types.StatusInformer{}, false
+}
+
+func (h *daemonSetEventHandler) cast(obj interface{}) *appsv1.DaemonSet {
+	r, _ := obj.(*appsv1.DaemonSet)
+	return r
+}
+
+// The pods in a daemonset can be identified by the match label set in the daemonset and the
+// "controller-revision-hash" can be used to determine if they are all the in the same daemonset
+// version.
+func (h *daemonSetEventHandler) calculateDaemonSetState(r *appsv1.DaemonSet) types.State {
+	if r == nil {
+		return types.StateUnavailable
+	}
+
+	selector := ""
+	for key, value := range r.Spec.Selector.MatchLabels {
+		if len(selector) > 0 {
+			selector += ","
+		}
+		selector += fmt.Sprintf("%s=%s", key, value)
+	}
+
+	// Gather all pods by looping until server responds with an empty continue field.
+	pods := make([]corev1.Pod, 0)
+	for {
+		result, err := h.clientset.CoreV1().Pods(h.targetNamespace).List(context.TODO(), metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			log.Printf("failed to get daemonset pod list: %s", err)
+			return types.StateUnavailable
+		}
+		pods = append(pods, result.Items...)
+
+		if len(result.Continue) == 0 {
+			break
+		}
+	}
+
+	// If the pod version labels are not all the same, then the daemonset is updating.
+	currentVersion := ""
+	for _, pod := range pods {
+		version, exists := pod.Labels[DaemonSetPodVersionLabel]
+		if !exists {
```

Comment:
does this label exist in older kubernetes versions like 1.19?

Reply:
it looks like [this is the commit](https://github.com/kubernetes/kubernetes/commit/25a67fc5cdac9c3b41bd0c7fdfb0a2c24a925894) that introduced the "controller-revision-hash" annotation to daemon sets in 2018 and the [first release](https://github.com/kubernetes/kubernetes/releases/tag/v1.19.0) of k8s v1.19 was in august 2020, so I'm pretty sure it's in all the versions we support.